### ISSUE TITLE
record export with uninformative data

### ISSUE THREAD

**Issue by berryld1** [vubiostat/redcapAPI#293]
I have observed this behavior. 

Say I call:

exportRecordsTyped(
  rcon = rcon,
  fields = c("record_id", "field_1", "field_2"))

Any records without a value in  either field_1 or field_2 -- which appears to include, for checkbox fields, value = 0 -- do not get returned on this call.

Or, at least, that's what I think I've observed. 

Does that seem right?

Maybe this is a side effect of trying not to return empty repeating instances?

HOWEVER, this can be problematic and misleading when specifying a set of fields -- just because there's no data in those fields (or checkbox value = 0) doesn't mean that record doesn't exist. 

Checkbox value = 0 can sometimes be meaningless - but not necessarily.

Here's the example I just encountered: I need to run a database transformation on a redcap I've populated with legacy data, processed and transformed to the redcap structure. As is often the case with such data, post-hoc, all of a sudden the client pipes up with some new data definitions for the legacy data, such that things need to be adjusted. I need only fields related to whether or not the tissue has been distributed, which all branch from an initial checkbox variable, call it distributed, where distributed___1 = 1 means distributed, = 0, no.

"Distributed" in the legacy data turns out to not actually mean distributed, so what I decided to do was to move all "distribution" data to a "legacy distribution" text field, so as to update to "not distributed", but not lose any of this legacy "distribution" data. For records with a "distribution", whatever data available will go in this "legacy distribution" field; for records without legacy distribution, I want to put "NONE" in the legacy distribution field.

The problem is, if I call just this set of fields I need, then subset on distributed___1 = 1 vs. 0, I get no records in the "0" subset. To get these records, I have to know (and remember) to explicitly include a field that I know will have data in it.

Which is fine, if someone knows and remembers to do this, or is looking carefully at what is happening when a script runs - but if not, these records would end up being passed over - without warning or otherwise.

**Comment by nutterb** [comment]
Do you get the desired data if you add `filter_empty_rows = FALSE` to your call?

I'm not sure if you will or won't, and don't have the resources with me to set up a test environment.

**Comment by berryld1** [comment]
Thanks, Benjamin!
No, filter_empty_rows = F doesn't seem to help - still get 0 rows data.
If I add a field that I know has data in it, get ~5000 rows of data

**Comment by nutterb** [comment]
I have nothing constructive to add at the moment. But this has my curiosity picqued. I'll look at this during my break.

**Comment by spgarbet** [comment]
The "0" mapped checkbox problem rears it's head again in a new form. This thing is a hydra.

**Comment by nutterb** [comment]
Ohhh...I completely failed to comprehend this was a checkbox___0 thing. This just got even more fun!

**Comment by berryld1** [comment]
Ah! Sorry. I explained poorly. The word "value" is ambiguous in the context of a checkbox! 

The checkbox in question here only has one response option, coded 1 - just that, at the moment, for a certain set of records, checkbox___1 = 0 (so, that's what I mean by value 0 in this context), and that same set of records has no values in any of the other variables of interest that I'm requesting.

So... it's a different flavor of fun, which appears to be, in dropping rows with no data (if that's what's happening?), redcapAPI is considering a 0 value for a checkbox - with the 1/0, TRUE/FALSE meaning of "value" here, i.e., the value of checkbox___1 = 0- to be non-data. 

Which is an interesting question, because, to a certain degree this is true - checkboxes will always take the default value of FALSE - even if, say, they are subject to branching logic, and are not called by that branching logic, so really should have no value at all.

Benjamin - i was certain your suggestion of filter_empty_rows = F would solve the problem, and then I would have to eat humble pie for submitting an issue instead of reading the documentation ;)

Humble pie still warranted (because I should have read the documentation and already tried that!) - but filter empty rows = F didn't fix it - only including a variable in the "fields" argument that I know has a non-missing, non-checkbox-0 (FALSE) value for all records.

**Comment by nutterb** [comment]
Well, we're not crazy!  I just used the REDCap API Playground. First thing I did was pull the data for a field where none of the checkboxes are checked.  I got this back. 

![image](https://github.com/vubiostat/redcapAPI/assets/1848445/b6e36449-2aa3-4670-90a3-30d9f246478a)

If I check a single box in one record, instead, I get this

![image](https://github.com/vubiostat/redcapAPI/assets/1848445/649fea61-70cc-401d-8495-4af9abc60f1c)

Now I uncheck the box, and include a field that has a value, and I get this

![image](https://github.com/vubiostat/redcapAPI/assets/1848445/ad327484-1f39-4cf2-843c-ed4c2c7404ae)

One last test, I make the value of mc_1 to be a missing value (it's a dropdown, for the record).  And BOOM! it's all missing values again.

![image](https://github.com/vubiostat/redcapAPI/assets/1848445/ae854c3a-8c7d-43b9-a59d-b2a66daa796e)

**Diagnosis**

We have a "Truthiness" problem. It would seem that REDCap pushes out the data for a record when the record has at least one "Truthy" field. Truthy, in this sense, would mean something that implies the existence of a thing. On the other hand, things like NULL, NA, empty strings, and 0 are "falsy" and imply the absence of a thing. 

That's rather inconvenient for us as far as checkbox fields go. But I think I can work around this. I think it's safe to say to a checkbox field that presents as an empty space may be considered a 0.

**Comment by spgarbet** [comment]
Wow! This explains so much and puts our library into a bind on removing empty rows. I.e., the API is inconsistent in this regard and we can't have it both ways. Let me elucidate the problem with an example:

Solution Idea: Post retrieval of data (pre-empty row filter), mark all zero coded fields that are blank as "0" in the raw. This would result in every form that contained a zero coded field never having "emptiness". I.e., zero-coded fields would always have a value whether or not a user entered a value. I.e., data would be "made-up" because of this. 

2nd Solution Idea: Filter empty rows and then mark all zero-coded fields that are blank as "0" in the raw. This would result in a data point that only contained a single zero coded field entered by the user being excluded from the data. 

Thus in the general case, a zero-coded field in a REDCap project has a choice between "making up data" or "potential data loss". This is due to the interaction of 2 things: (1) REDCap has no means to mark a form as having entries or being empty in the export and (2) REDCap treating "0" as not an entry, but allowing for zero coded fields. 

There is no proper answer. I think our warning message about zero-coded fields needs to be LOUDER.

**Comment by spgarbet** [comment]
There is one possible out to try and it would require a lot of work. The JSON export is said to not report rows that are empty. Try your experiment with JSON and see what happens.

**Comment by nutterb** [comment]
No change with the JSON.  We'd have the same dilemma. 

![image](https://github.com/vubiostat/redcapAPI/assets/1848445/b5e67426-dad0-465f-a837-33f39104c73b)

I'm kind of on board with your first solution idea.  With respect to checkbox fields (and only with checkbox fields), I'm not terribly bothered with "making up" 0 data. Afterall, the default value of any checkbox is 0 until a user makes an explicit action to change that. 

Even with your concern about forms, when I set the `form_complete` dropdown to "Complete", these results don't change. That's probably the closest proxy we have to if a user intended the absence of markings. 

I also don't think there is any additional detriment when we encounter zero-coded checkbox fields.  if `checkbox___0` is a missing value, it's safe to assume that is a 0 (Unchecked) in the raw data. All of the problems that apply thereafter are still applicable, but we've not made it worse. 

In any case, the key is that the missing value has to become a zero in the RAW data, before we do any data processing. You'll see a pull request come through shortly.  It might alleviate some of your concerns.  Or intensify them. 

To sum up my feelings

1. I think for checkboxes, we are probably okay inserting 0s as a replacement for missing values, since the missing values only appear in the absence of truthy data.
2. The fact that all three of us on this thread were surprised at this behavior suggests that replacing with 0 is probably what a lot of people are expecting anyway.

**Comment by spgarbet** [comment]
The proposal is to use the form_complete and to then treat the raw as "0" for this case. That might work. It still runs the risk of "making up" data, but if the form is marked complete this assumes "0" is the default always.

**Comment by spgarbet** [comment]
Here's what I think we should do: (1) Use the sandbox project and create a very simple example with a zero coded field. Then create a simple demo of the indeterminancy between "not entered" and "0" coded. Open a support request that the core REDCap team fix it at the server level. I.e., if it's a "0" report it as a "0".

**Comment by nutterb** [comment]
Sounds great. If you can't find anyone with an objection, I'll demolish the sandbox tomorrow so we can have a clean project file to send them.

**Comment by berryld1** [comment]
Wow! Great findings!!! I'm very interested to hear redcap's response to this - there are a lot of inherent tensions in how checkboxes work in redcap (or, maybe, everywhere).
Your results, Benjamin, make me think, checkboxes *never* take a 0 value in the actual back-end database. They are either 1 (checked) or "" (unchecked).
When redcap deems that data should be returned for a "row", at that point, all unchecked checkboxes are filled with 0s in the return data - even checkbox fields that, in that dataset, would be hidden by branching logic, so really *shouldn't* take a value.
When redcap deems no data should be returned for a "row", this filling doesn't happen. 
Things that make me think this:
- Benjamin's testing results here
- In a returned dataset with data (meaning: other data in the row), all unchecked checkboxes get 0 - even if that field is hidden by branching logic in the GUI, so no value should be expected (this kind of drives me nuts).
- In the case of a checkbox field with only one option (I use these sometimes for things like: check here to enter another finding, so the checkbox is really just driving branching logic or another particular behavior, and I'm not interested in an alternative check), if that field is made required, and the box is left unchecked, the GUI will object. 
This is consistent with the checkbox-dilemma - i.e., did the user mean no/0 to that box, or forgot to check it? - I think this is what you mean by truthy/falsey, Benjamin?
BUT if you have checkboxes, that's a risk you've decided to take, and "unchecked" should be represented as truthy (imho).

**Comment by nutterb** [comment]
yeah, you've got the gist of truthy/falsy. If you like procrastinating as much as I do, this is a decent introduction to it. https://developer.mozilla.org/en-US/docs/Glossary/Truthy

R doesn't have a native concept of truthy.  The closest analog is `isTRUE(as.logical(x))`. And shiny has its own implementation of truthiness that, oddly, considers 0 to be truthy. 

I'm not entirely sure what gets stored in the backend.  It could be that the zeros are stored, but during export a value is only given if there is at least one truthy value on the row. It could also be stored as "" and converted to zero in the presence of at least one truthy value.  My limited-information guess is that whatever is stored was chosen to minimize storage space on the disk. 

Regardless of these mechanics, you're analysis highlights how difficult it is to tell the difference between a meaningful-unchecked box and a missing-value-unchecked box.

**Comment by spgarbet** [comment]
I'm lost in a stew of zero coded truthy in my head.

What if I'm overthinking things? Could we just change the castCheckbox to treat "" as "0"?

**Comment by nutterb** [comment]
I went playing in the sandbox, and picked up a couple of extra tidbits. The highlights:

We see some weird behavior because the API allows us to select fields in a way that is not available through the UI's data export page. Users who download data from the UI will never encounter this behavior.

We seem to be correct that the checkboxes only get converted to 0 so long as there is at least one truthy value in the row. And since the UI _always_ exports the record ID, and since the record ID is _always_ truthy, the UI data exports always have their checkboxes converted to 0. 

Throughout the discussion, I've been able to see various approaches to handling this oddity, and there is no approach that doesn't come with disadvantages. No matter what we do, we are going to have some nuanced situation where the way we treat this is inconvenient. 

With that in mind, I'm firmly convinced that the "best" solution is to make our raw data export as similar to the UI data export as we possibly can. This means converting blanks in checkboxes to zero prior to filtering out empty rows. And while this has drawbacks, at least it has the same drawbacks a user exporting data through the UI would be subject to.

I will add a statement to documentation to address this and reopen the pull request.

**Comment by berryld1** [comment]
Thanks, Shawn and Benjamin! Shawn, please remind me how to update the package from github? I never can remember how--nor remember where I documented how ;)

**Comment by berryld1** [comment]
p.s. agree with you, Benjamin, that mirroring what happens in the GUI seems like the minimum-damage solution!

**Comment by nutterb** [comment]
@berryld1 `devtools::install_github("vubiostat/redcapAPI")`

**Comment by spgarbet** [comment]
@berryld1 did this work for you?

**Comment by berryld1** [comment]
Sorry to be so long to get back! Yes, this fix works for my case!

**Comment by spgarbet** [comment]
Yahoo!!!
